### ISSUE TITLE
LoggingFactory race condition

### ISSUE THREAD

**Issue by AFulgens** [dropwizard/dropwizard#1205]
Context: I inherited a maven parent pom which had parallel testing setup in surefire with
```
<parallel>classes</parallel>
<threadCount>2</threadCount>
<perCoreThreadCount>true</perCoreThreadCount>
```

Issue: I have multiple test classes in a JUnit test suite which use `ResourceTestRule`s and/or `DropwizardClientRule`s. These are `@ClassRule`s which have to be `public static`, thus they are initialized at class load time.

They internally call `LoggingFactory.bootstrap()`, which ends up in a race condition because of the parallelization over the test classes, which in the end will lead to a `ClassCastException` at `io.dropwizard.logging.LoggingFactory.bootstrap(LoggingFactory.java:43)`. The exception is because the SLF4J `org.slf4j.helpers.SubstituteLogger` cannot be cast to `ch.qos.logback.classic.Logger`.

Proposed fix: If I understand correctly any logging calls to the `SubstituteLogger` are anyways dropped by the logger framework, thus encapsulating the cast at `LoggingFactory.java:43` into a try-catch and doing nothing with the exception would solve the issue - maybe emitting a warning that this has happened during init, but that warning should be anyways emitted by SLF4 ("The following set of substitute loggers may have been accessed [...]"). No idea how to write tests for this though.

**Comment by mbknor** [comment]
We have also seen this issue when bootstrapping dropwizard apps instrumented with newrelic. This sometimes (always?) causes slf4j/logback to init slower than usuall, which causes this problem. I have a proposed fix which I will send as a pull-request soon.

**Comment by mbknor** [comment]
Looks like this issue has already been fixed in v 0.9.0-rc5 - https://github.com/dropwizard/dropwizard/commit/847275afd08b217bd98c633c6c7c2d1ab84b8cc0

**Comment by mrmanc** [comment]
@mbknor that’s awesome news… thanks for the update. We’ll look forward to the release. :+1:

**Comment by AFulgens** [comment]
@mbknor Awesome, thanks for the update, looking forward for v0.9.0 :clap:

**Comment by nickbabcock** [comment]
Looks like this issue is resolved, so I will close this. If the issue persists, reply back and I can re-open the issue.

**Comment by arteam** [comment]
It actually was fixed in 14bbdad in the efforts to run the Dropwizard tests in parallel. It took some time to get it correctly, but now the build looks stable enough.

**Comment by mbknor** [comment]
nice. For the record, here is the workaround we have created for fixing this issue related to newrelic and dropwizard 0.7 and 0.8 in production.

https://gist.github.com/mbknor/34944ea4589a5fc6974c
